### ISSUE TITLE
Can't contact support

### ISSUE THREAD

**Issue by JohnLouderback** [trueinteractions/tint2#106]
Hi Trevor, 

For some reason I can't send mail to your support address. It's getting bounced back to me. Here's my email I tried sending to support:

"Hey Trevor,

How are you doing? I wanted to make sure you saw my comment here: https://groups.google.com/forum/#!category-topic/trueinteractions/WCn8fx5F2hQ
Being able to use Chromium is a big priority. I know you said you were going to prioritize transparency, but it dawned on me that these two features may interact. Chromium with transparency is a big concern for the project I'm working on. Another point of interest for me if the difficulty of working between the node and browser contexts. I don't know what the technical limitations are here, but it would be nice to pass data between contexts besides strings, such as object literals. Being able to pass callbacks becomes very difficult otherwise. Another alternative could be the ability to access a webview's window object from the node context. I understand the issue of App Stores and I'm not sure if there are other technical limitations involved, but if it's possible, perhaps even allowing for shared contexts via a switch, if App Store compatibility is not a concern. I might be inclined to finish this project in Tint v1, but the ability to use the bridge API of Tint v2 is far too tempting. Anyway, thank you for your time. Just throwing out some concerns and ideas.

Thanks,
-John"

**Comment by trevorlinton** [comment]
Hi John,

We have been getting your emails (and responding).  We received:

One from Jul 29th, 2014 and responded August 4th 2014 "A few questions"

Otherwise most of the communications have been through github.  We're looking into making libchromecontent available as a plugin to Tint to use optionally, this would give debugging and transparency support as you've noted above. 

As for making node available in the security javascript context of a webview that's a difficult problem.  The issue surrounds things that require more than a technical solution.  Apple has been fairly adamant about being cautious at revealing backend or escalated privileges to a webview context, simply because its a large security risk due to the nature of HTTP/HTML and Javascript. They've even gone to extents of banning some phone gap plugins that expose too much capabilities in a web view.

Imagine if you had node available in the browser and an iframe or other javascript source was accidentally included remotely (or perhaps some user generated content allowed javascript to execute); this would give someone full access to other peoples computers.  

To remedy this we purposely made it more difficult to execute node-based javascript in a browser context (or to put it another way, made a very clear distinction of what can and can't execute privileged functions) with tint2.  

We can most likely add the feature to be able to access the Window context from the node end (however data passed from function calls would still require serialization, e.g., no native objects can be passed back and forth, only copiable data and function callbacks). Do you feel that would fix the headaches?

**Comment by JohnLouderback** [comment]
Hi Trevor,

I sent an email on January 12th and the one in this post on the 22nd. Did you receive those? It could be something strange on my end, but I wanted to let you know, just in case you're missing emails from others as well.

The plugin solution for chromium sounds excellent! It sounds like it will be tightly integrated if it will support transparency. I'm excited for that.

I definitely see where you're coming from on the node in the webview thing. I think my most major hang up is the confusion of writing UI code both on the node side and the webview side. For instance, if I had code to build a menu that built the menu using the native menu bar on OS X, but used the DOM to build a menu on Windows. These menu items would of course need event handlers and such, which should be the same on both ends. The trouble arises here when I need to decide where this UI code lives. If I put it on either side, binding the event handler callbacks on the other side become unwieldy, if even possible. Perhaps I have the wrong perspective here. I'm hoping there's currently a pattern I can employ here to meet this use case. I understand in the vast majority of situations, using the postMessage model is extremely similar to client/server model, which works great until the complication of managing two UIs come into play. What are your thoughts?

All in all, I have to say that Tint 2 shows a ton of promise and I'm excited to begin migrating towards it. The API for tapping into the native libraries is an awesome feature sorely missed from any of the competition that I am aware of. Once it has the chromium plugin, I think development of native applications will be unparalleled.

**Comment by trevorlinton** [comment]
@JohnLouderback 

Hi John,  we did receive those, i'm unsure some of them maybe are getting stuck in our spam folders, some of the responses may have came from trevor.linton at gmail dot com.  Can you add that to your white list, we'll take a look at our email systems as well.

Also, we took a look into integrating chromium and it seems pretty straight forward. CEF (chromium embedded) provide .NET Windows DLL's making the process all that easier. 

We put together a fairly straight forward example that you can download and try out on your machine at: https://www.trueinteractions.com/tint-chromium.zip

We'll publish a module "tint-chromium" (?) on NPM with the following features:

1. The latest version of chromium with an auto-update CI build process to push out new versions.
2. A fully compatible JS WebView API for the module (see https://www.trueinteractions.com/tint2/docs/WebView.html)
2. Ability to open a debugging port on the WebView (and the debugging window).
3. Transparency support
4. Hardware acceleration support
5. Packaging/build support

We should have the Chromium node module on NPM for Monday (March 30th) and the 2.1.6 release. Did we capture all of the requirements?  Let me know if there's anything in addition you'd like to see.

**Comment by JohnLouderback** [comment]
I'm seeing this when running the that code through the latest Tint:
```
Bridge.js:288
  var types = dotnet.loadAssembly(assembly);
                     ^
Error: System.IO.FileLoadException Could not load file or assembly 'file:///D:\Users\John\Downloads\tint-chromium\CefSharp.dll' or one of its dependencies. Operation is not supported. (Exception from HRESULT: 0x80131515)
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at CLR.LoadAssembly(FunctionCallbackInfo<v8::Value>* args)
    at Error (native)
    at importOnto (Bridge.js:288:22)
    at Object.imports (Bridge.js:300:5)
    at Object.<anonymous> (D:\Users\John\Downloads\tint-chromium\index.js:3:9)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
```

I think those meet all the requirements, but maybe I'm assuming too much. Will this be available on both Windows and OS X, or just Windows? Will transparency support also support "click throughs" so that fully transparent areas do not receive clicks and that windows and controls behind the transparent webview can receive clicks instead? With this, will I also see window transparency in Windows? Also, I don't know if this is out-of-scope for this request, but an issue I had with Tint v1 previously, was that drop shadows were not transparent to clicks on transparent windows. Is it possible to support click throughs on portions of the webview that do not current receive click events normally? If it supports the features of the latest Chrome, what you mentioned, and potentially what I've brought up, I can't think of anything else at the moment. Chromium in and of itself is a huge addition and feature. Thanks a lot for all of your attention here Trevor, this is a great product. It's great to see it expand. I didn't fully realize the power until I saw how you easy incorporated Chromium through the DLLs like it was no big deal.

**Comment by JohnLouderback** [comment]
On the first half of that, it appears the issue is just because Windows blocks DLLs initially if they're from downloaded zip archives. Anyway, I'm unblocking them now. Everything in that demo should work well. I'll let you know if that isn't the case.

**Comment by trevorlinton** [comment]
@JohnLouderback 

There's options on the Cef CefSharp and ChromiumWebBrowser classes to initialize it with transparency, and according to the CEF folks it does support it but i'll dig a little deeper and see what we can find.

The app protocol shouldn't be a problem, CEF also supports built in app schemas, as for keeping it up with beta/canary, the latest CEF build is 39, unless i maintain my own upstream sources (which is harder than you'd think) we'd basically be pulling down whatever CEF has rolled out.  According to their site (https://github.com/cefsharp/CefSharp) CEF has v39 (which is about 3~4 months old, latest would be 42).  But not too shabby. I'd imagine the CEF team will keep it updated at more regular intervals (and have been in the past).

The webkit-app-region is supported in CEF as well so i'd imagine it would also work. 

There's a CEF Objective-C framework as well; objective-c frameworks are the same as .NET dll's, so we can pull down an OS X version as well.

Tint uses a CLR reflection bridge we developed in house for windows and one we co-developed with nodobjc for OSX/iOS. We're currently looking into the viability of using gobject on linux to support similar features.  These ffi/reflection bridges are incredibly powerful as they allow us to simply load up any existing .dll's or frameworks (or i suppose, .so g-objects on linux) and immediately have access to all of their native classes, enums, functions, etc.  The memory is all managed by v8 and auto released when the object is no longer needed. 

The Tint API (e.g., what you see in the docs) is just a "common denominator" framework, meaning it just standardizes the behavior across multiple operating systems.   On any Tint object you create you can just access .native to grab the underlying native object.

For example, if you do 
 
```javascript
require('Common'); 
var win = new Window(); 
console.log(window.native);
```

On Windows it'll say its a System.Windows.Window class which is this object here: https://msdn.microsoft.com/en-us/library/system.windows.window(v=vs.110).aspx

On OS X it'll say its a NSWindow class which is this class here:
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ApplicationKit/Classes/NSWindow_Class/index.html

This makes it easier to make cross-OS applications as you have a common API that you can use that you know works the same across different operating systems, but if you really need to you can extract the native object and start playing around with it. 

Or, you can even create NPM modules that combine a bunch of native functionality (and standardize it) to make your components even that more usable. 

The net effect is its allowed us to do some leapfrogs over what previously with node-webkit/atomshell required a 2 hour build of chromium on any small change... and required including an 80mb runtime.

Hope this helps, we'll keep plugging away on the chromium module but I don't see any problems with any of the requests above, should be easy to add.

I'll update you on monday on our status.

**Comment by JohnLouderback** [comment]
Hi Trevor,

That sounds awesome. Eager to hear from you tomorrow. I did want to add, in my quick tests with CEF Sharp, the ``-webkit-app-region: drag`` does not appear to work. At least it doesn't out of the box. I don't know if more needs to be done to enable it, but I figured I'd call that out.

**Comment by trevorlinton** [comment]
@JohnLouderback Hi John, we've created a new repo https://github.com/trueinteractions/tint-chromium for issues related to the chromium plugin in development.

In addition, you can install the chromium module using the instructions on the repo (e.g., ```npm install```) if you have issues with the module feel free to file issues on that repo and we'll address them there (makes it easier for us to manage).

In the meantime i'm going to close out this thread on the main repo.

- Trevor
